drivers/i2c/i2c_nxp_ii2c.c

```diff
@@ -0,0 +1,416 @@
+{
+	I2C_Type *base = get_base(dev);
+	struct nxp_ii2c_data *data = dev->data;
+	i2c_master_transfer_t transfer;
+	status_t status;
+	struct i2c_msg *msg = &data->msgs[data->msg];
+
+	if (I2C_MSG_ADDR_10_BITS & msg->flags) {
+		nxp_ii2c_async_done(dev, data, -ENOTSUP);
+		return;
+	}
+
+	/* Initialize the transfer descriptor */
+	transfer.flags = nxp_ii2c_convert_flags(msg->flags);
+	transfer.slaveAddress = data->addr;
+	transfer.direction = (msg->flags & I2C_MSG_READ) ? kI2C_Read : kI2C_Write;
+	transfer.subaddress = 0;
+	transfer.subaddressSize = 0;
+	transfer.data = msg->buf;
+	transfer.dataSize = msg->len;
+
+	/* Prevent the controller to send a start condition between
+	 * messages, except if explicitly requested.
+	 */
+	if (data->msg != 0 && !(msg->flags & I2C_MSG_RESTART)) {
+		transfer.flags |= kI2C_TransferNoStartFlag;
+	}
+
+	/* Start the transfer */
+	status = I2C_MasterTransferNonBlocking(base, &data->handle, &transfer);
+
+	/* Return an error if the transfer didn't start successfully
+	 * e.g., if the bus was busy
+	 */
+	if (status != kStatus_Success) {
+		I2C_MasterTransferAbort(base, &data->handle);
+		nxp_ii2c_async_done(dev, data, -EIO);
+	}
+}
+
+static int nxp_ii2c_transfer_cb(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+				   uint16_t addr, i2c_callback_t cb, void *userdata)
+{
+	struct nxp_ii2c_data *data = dev->data;
+
+	int res = k_sem_take(&data->lock, K_NO_WAIT);
+
+	if (res != 0) {
+		return -EWOULDBLOCK;
+	}
+
+	data->msg = 0;
+	data->msgs = msgs;
+	data->num_msgs = num_msgs;
+	data->addr = addr;
+	data->cb = cb;
+	data->userdata = userdata;
+	data->addr = addr;
+
+	nxp_ii2c_async_iter(dev);
+
+	return 0;
+}
+
+#endif /* CONFIG_I2C_CALLBACK */
+
+static void nxp_ii2c_isr(const struct device *dev)
+{
+	I2C_Type *base = get_base(dev);
+	struct nxp_ii2c_data *data = dev->data;
+
+	I2C_MasterTransferHandleIRQ(base, &data->handle);
+}
+
+static int nxp_ii2c_init(const struct device *dev)
+{
+	I2C_Type *base;
+	const struct nxp_ii2c_config *config = dev->config;
+	struct nxp_ii2c_data *data = dev->data;
+	uint32_t clock_freq, bitrate_cfg;
+	i2c_master_config_t master_config;
+	int error;
+
+	DEVICE_MMIO_NAMED_MAP(dev, reg_base, K_MEM_CACHE_NONE | K_MEM_DIRECT_MAP);
+
+	if (!device_is_ready(config->clock_dev)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	if (clock_control_get_rate(config->clock_dev, config->clock_subsys,
+				   &clock_freq)) {
+		return -EINVAL;
+	}
+
+	error = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (error) {
+		return error;
+	}
+
+	base = get_base(dev);
+
+	k_sem_init(&data->lock, 1, 1);
+	k_sem_init(&data->device_sync_sem, 0, K_SEM_MAX_LIMIT);
+
+	I2C_MasterGetDefaultConfig(&master_config);
+	I2C_MasterInit(base, &master_config, clock_freq);
+	I2C_MasterTransferCreateHandle(base, &data->handle,
+				       nxp_ii2c_master_transfer_callback, (void *)dev);
+
+	bitrate_cfg = i2c_map_dt_bitrate(config->bitrate);
+
+
```

Comment:
You can delete some of these blank lines

Reply:
Done, thanks.